pageserver/src/consumption_metrics/upload.rs

```diff
@@ -107,6 +104,62 @@ fn serialize_in_chunks<'a>(
     })
 }
 
+trait RawMetricExt {
+    fn as_event(&self, node_id: &str) -> Event<Ids, Name>;
+    fn update_in_place(&self, event: &mut Event<Ids, Name>, node_id: &str);
+}
+
+impl RawMetricExt for RawMetric {
+    fn as_event(&self, node_id: &str) -> Event<Ids, Name> {
+        let MetricsKey {
+            metric,
+            tenant_id,
+            timeline_id,
+        } = self.0;
+
+        let (kind, value) = self.1;
+
+        Event {
+            kind,
+            metric,
+            idempotency_key: idempotency_key(&node_id),
+            value,
+            extra: Ids {
+                tenant_id,
+                timeline_id,
+            },
+        }
+    }
+
+    fn update_in_place(&self, event: &mut Event<Ids, Name>, node_id: &str) {
+        use std::fmt::Write;
+
+        let MetricsKey {
+            metric,
+            tenant_id,
+            timeline_id,
+        } = self.0;
+
+        let (kind, value) = self.1;
+
+        *event = Event {
+            kind,
+            metric,
+            idempotency_key: {
+                event.idempotency_key.clear();
+                let next = IdempotencyKey::generate(node_id);
+                write!(event.idempotency_key, "{}", next).unwrap();
```

Comment:
This can be `event.idempotency_key = format!("{next}")`. Also saves the `.clear()` call above.

Reply:
That would allocate again.